prov/tcp/src/tcpx_conn_mgr.c

```diff
@@ -178,48 +180,108 @@ static int handle_poll_list(struct poll_fd_mgr *poll_mgr)
 	socklen_t len;
+	uint16_t cm_data_sz;
 	int ret, status;
 
 	assert(poll_info->fid->fclass == FI_CLASS_EP);
 	ep = container_of(poll_info->fid, struct tcpx_ep, util_ep.ep_fid.fid);
 
-	len = sizeof(status);
-	ret = getsockopt(ep->conn_fd, SOL_SOCKET, SO_ERROR, &status, &len);
-	if (ret < 0 || status) {
-		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL, "connection failure\n");
+	if (ESTABLISH_CONN == poll_info->state) {
+		assert(poll_mgr->poll_fds[index].revents == POLLOUT);
 
-		memset(&err_entry, 0, sizeof err_entry);
-		err_entry.fid = poll_info->fid;
-		err_entry.context = poll_info->fid->context;
-		err_entry.err = status ? status : errno;
+		len = sizeof(status);
+		ret = getsockopt(ep->conn_fd, SOL_SOCKET, SO_ERROR, &status, &len);
+		if (ret < 0 || status) {
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL, "connection failure\n");
+			ret = status? status : errno;
+			goto err;
+		}
 
-		ret = fi_eq_write(&ep->util_ep.eq->eq_fid, FI_SHUTDOWN,
-				  &err_entry, sizeof(err_entry), UTIL_FLAG_ERROR);
-	} else {
-		memset(&cm_entry, 0, sizeof cm_entry);
-		cm_entry.fid = poll_info->fid;
+		conn_req.type = TCPX_CONN_REQ;
+		conn_req.cm_data_sz = htons(poll_info->cm_data_sz);
 
-		ret = fi_eq_write(&ep->util_ep.eq->eq_fid, FI_CONNECTED,
-				  &cm_entry, sizeof(cm_entry), 0);
+		ret = tcpx_cm_send(ep->conn_fd, &conn_req, sizeof(conn_req));
+		if (ret)
+			goto err;
+
+		if (conn_req.cm_data_sz) {
+			ret = tcpx_cm_send(ep->conn_fd, poll_info->cm_data,
+					   poll_info->cm_data_sz);
+			if (ret)
+				goto err;
+		}
+
+		poll_info->state = RCV_RESP;
+		poll_mgr->poll_fds[index].events = POLLIN;
+		return 1;
 	}
 
-	if (ret < 0)
-		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL, "Error writing to EQ\n");
+	if (RCV_RESP == poll_info->state) {
+
+		assert(poll_mgr->poll_fds[index].revents == POLLIN);
+		ret = tcpx_cm_recv(ep->conn_fd, &conn_resp, sizeof(conn_resp));
+		if (ret)
+			goto err;
+
+		cm_data_sz = ntohs(conn_resp.cm_data_sz);
+		if (cm_data_sz) {
+			if (cm_data_sz > TCPX_MAX_CM_DATA_SIZE)
+				goto err;
+
+			ret = tcpx_cm_recv(ep->conn_fd, poll_info->cm_data,
+					   cm_data_sz);
+			if (ret)
+				goto err;
+		}
+
+		cm_entry = calloc(1, sizeof(*cm_entry) + cm_data_sz);
+		if (!cm_entry) {
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL, "mem alloc failed\n");
+			ret = -FI_ENOMEM;
+			goto err;
+		}
+
+		cm_entry->fid = poll_info->fid;
+		memcpy(cm_entry->data, poll_info->cm_data, cm_data_sz);
+
+		ret = fi_eq_write(&ep->util_ep.eq->eq_fid, FI_CONNECTED,
+				  cm_entry, sizeof(*cm_entry) + cm_data_sz, 0);
+		if (ret < 0) {
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL, "Error writing to EQ\n");
+			goto remove_fd;
+		}
+		goto remove_fd;
+	}
+	ret = -FI_EINVAL;
+err:
+	memset(&err_entry, 0, sizeof err_entry);
+	err_entry.fid = poll_info->fid;
+	err_entry.context = poll_info->fid->context;
+	err_entry.err = ret;
+
+	fi_eq_write(&ep->util_ep.eq->eq_fid, FI_SHUTDOWN,
+		    &err_entry, sizeof(err_entry), UTIL_FLAG_ERROR);
+remove_fd:
+	return 0;
```

Comment:
This function is getting long.  Consider breaking into subcalls.  May also want to convert if() state checks into switch statement, so we can trap and handle unexpected states.